mmvec/heatmap.py

```diff
@@ -105,6 +107,134 @@ def ranks_heatmap(ranks, microbe_metadata=None, metabolite_metadata=None,
+        Set to -1 if not parsing semicolon-delimited taxonomies.
+    normalize: str
+        Column normalization strategy to use for heatmaps. Must
+        be "log10", "z_score", or None
+    color_palette: str
+        Color palette for heatmaps.
+    '''
+    if top_k_microbes is features is None:
+        raise ValueError('Must select features by name and/or use the '
+                         'top_k_microbes parameter to select features to '
+                         'include in the heatmap.')
+
+    # validate microbes
+    if features is not None:
+        microbe_ids = set(microbes_table.ids('observation'))
+        missing_microbes = set(features) - microbe_ids
+        if len(missing_microbes) > 0:
+            raise ValueError('features must represent feature IDs in '
+                             'microbes_table. Missing microbe(s): {0}'.format(
+                                missing_microbes))
+    else:
+        features = []
+
+    microbes_table = microbes_table.to_dataframe().T
+    metabolites_table = metabolites_table.to_dataframe().T
+
+    # optionally normalize tables
+    if normalize != 'None':
+        microbes_table = _normalize_table(microbes_table, normalize)
+        metabolites_table = _normalize_table(metabolites_table, normalize)
+        cbar_label = normalize + ' Frequency'
+    else:
+        cbar_label = 'Frequency'
+
+    # find top k microbes (highest relative abundances)
+    if top_k_microbes is not None:
+        # select top relative abundances
+        top_microbes = microbes_table.apply(
+            lambda x: x / x.sum(), axis=1).sum().sort_values(ascending=False)
+        # TODO: add option for selecting top_k_microbes by rank
+        # top_microbes = ranks.max(axis=1).sort_values(ascending=False)
+        top_microbes = top_microbes[:top_k_microbes].index
+        # merge top k microbes with selected features
+        # use list comprehension instead of casting as set to preserve order.
+        features = features + [f for f in top_microbes if f not in features]
+
+    # select samples in which microbes are most abundant feature
+    if keep_top_samples:
+        select_microbes = microbes_table[microbes_table.apply(
+            pd.Series.idxmax, axis=1).isin(features)]
+
+    # filter select microbes from microbe table and sort by abundance
+    sort_orders = [False] + [True] * (len(features) - 1)
+    select_microbes = select_microbes[features]
+    select_microbes = select_microbes.sort_values(
+        features, ascending=sort_orders)
+
+    # find top K metabolites (highest positive ranks)
+    if top_k_metabolites != 'all':
+        microb_ranks = ranks.loc[features]
+        top_metabolites = microb_ranks.max()
+        top_metabolites = top_metabolites.sort_values(ascending=False)
+        top_metabolites = top_metabolites[:top_k_metabolites].index
+        # grab top K metabolites in metabolite table
+        select_metabolites = metabolites_table[top_metabolites]
```

Comment:
this looks funky - this should be top K metabolites for each of the microbes specified.

Reply:
This block should be something like

```
top_metabolites = []
for m in select_microbes.columns:
    top_metabolites += list(ranks.loc[m].sort_values().index[-top_k_metabolites:])
select_metabolites = metabolites_table[top_metabolites]
```
Note that there will be redundant metabolites (since some microbes can have the same top co-occurring metabolites).  But I'm ok to just acknowledge this for now and have a fix later.